include/libtorrent/packet_pool.hpp

```diff
@@ -153,34 +158,36 @@ namespace libtorrent
 			if (p == nullptr) return;
 
 			packet_ptr pp(p); // takes ownership and may auto free if slab container does not move it
-			std::size_t allocated = pp->allocated;
+			int const allocated = pp->allocated;
 
-			if (allocated <= m_syn_slab.allocate_size) { m_syn_slab.try_push_back(pp); }
-			else if (allocated <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_push_back(pp); }
-			else if (allocated <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_push_back(pp); }
+			if (allocated == m_syn_slab.allocate_size) { m_syn_slab.try_push_back(pp); }
+			else if (allocated == m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_push_back(pp); }
```

Comment:
if I return a packet whose allocation is 529 into this slab, and next someone needs a packet that can fir 548 bytes, it will return an allocation that's *smaller* than the requested size. The only reason this code worked was because we would only return packets allocated with this same pool, which would have rounded up to 548, for small allocations.

What my patch does is to encode the fact that it was rounded up in the ``allocated`` member (which also allows the packet to grow without reallocating it). It also makes the freeing robust, since we don't make any assumptions about where the packet came from.

Am I misunderstanding something?